File: backend/core/src/modules/auth/auth.controller.ts
import { Body, Controller, Delete, Get, Param, Patch, Post, Req, UseGuards } from '@nestjs/common';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { RegisterDto } from './dto/register.dto';
import { RefreshDto } from './dto/refresh.dto';
import { VerifyEmailDto } from './dto/verify-email.dto';
import { ResendVerificationDto } from './dto/resend-verification.dto';
import { RequestPasswordResetDto } from './dto/request-password-reset.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { ChangePasswordDto } from './dto/change-password.dto';
import { RequestEmailChangeDto } from './dto/request-email-change.dto';
import { ConfirmEmailChangeDto } from './dto/confirm-email-change.dto';
import { UpdateProfileDto } from './dto/update-profile.dto';
import { DeleteAccountDto } from './dto/delete-account.dto';
import { JwtAuthGuard } from './jwt-auth.guard';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('register')
  register(@Body() dto: RegisterDto, @Req() req: any) {
    return this.authService.register(dto, this.extractSessionMeta(req));
  }

  @Post('login')
  login(@Body() dto: LoginDto, @Req() req: any) {
    return this.authService.login(dto, this.extractSessionMeta(req));
  }

  @Post('refresh')
  refresh(@Body() dto: RefreshDto) {
    return this.authService.refresh(dto.refreshToken);
  }

  @Post('verify-email')
  verifyEmail(@Body() dto: VerifyEmailDto) {
    return this.authService.verifyEmail(dto.token);
  }

  @Post('resend-verification')
  resendVerification(@Body() dto: ResendVerificationDto) {
    return this.authService.resendVerification(dto.email);
  }

  @Post('request-password-reset')
  requestPasswordReset(@Body() dto: RequestPasswordResetDto) {
    return this.authService.requestPasswordReset(dto.email);
  }

  @Post('reset-password')
  resetPassword(@Body() dto: ResetPasswordDto) {
    return this.authService.resetPassword(dto.token, dto.newPassword);
  }

  @Post('confirm-email-change')
  confirmEmailChange(@Body() dto: ConfirmEmailChangeDto) {
    return this.authService.confirmEmailChange(dto.token);
  }

  @UseGuards(JwtAuthGuard)
  @Post('logout')
  logout(@Req() req: any) {
    const user = req.user as { userId: string };
    return this.authService.logout(user.userId);
  }

  @UseGuards(JwtAuthGuard)
  @Get('me')
  me(@Req() req: any) {
    return this.authService.me(req.user?.userId);
  }

  @UseGuards(JwtAuthGuard)
  @Patch('me')
  updateMe(@Req() req: any, @Body() dto: UpdateProfileDto) {
    return this.authService.updateProfile(req.user?.userId, dto);
  }

  @UseGuards(JwtAuthGuard)
  @Get('sessions')
  sessions(@Req() req: any) {
    return this.authService.listSessions(req.user?.userId);
  }

  @UseGuards(JwtAuthGuard)
  @Delete('sessions/:id')
  revokeSession(@Req() req: any, @Param('id') id: string) {
    return this.authService.revokeSession(req.user?.userId, id);
  }

  @UseGuards(JwtAuthGuard)
  @Post('change-password')
  changePassword(@Req() req: any, @Body() dto: ChangePasswordDto) {
    return this.authService.changePassword(req.user?.userId, dto.currentPassword, dto.newPassword);
  }

  @UseGuards(JwtAuthGuard)
  @Post('request-email-change')
  requestEmailChange(@Req() req: any, @Body() dto: RequestEmailChangeDto) {
    return this.authService.requestEmailChange(req.user?.userId, dto.newEmail);
  }

  @UseGuards(JwtAuthGuard)
  @Post('delete-account')
  deleteAccount(@Req() req: any, @Body() dto: DeleteAccountDto) {
    return this.authService.deleteAccount(req.user?.userId, dto.password);
  }

  private extractSessionMeta(req: any) {
    const userAgent = typeof req?.headers?.['user-agent'] === 'string' ? (req.headers['user-agent'] as string) : undefined;
    const forwarded = typeof req?.headers?.['x-forwarded-for'] === 'string' ? (req.headers['x-forwarded-for'] as string) : undefined;
    const ipAddress = forwarded?.split(',')[0]?.trim() || req?.ip;
    return { userAgent, ipAddress };
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/controller/AuthController.java
package store.pjcloud.cowrite.core.controller;

import jakarta.servlet.http.HttpServletRequest;
import java.util.UUID;
import org.springframework.web.bind.annotation.*;
import store.pjcloud.cowrite.core.security.SecurityUtils;
import store.pjcloud.cowrite.core.service.AuthService;
import store.pjcloud.cowrite.core.service.AuthService.SessionInfo;
import store.pjcloud.cowrite.core.service.AuthService.SessionMeta;

@RestController
@RequestMapping("/auth")
public class AuthController {
    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @PostMapping("/register")
    public Object register(@RequestBody RegisterRequest request, HttpServletRequest httpRequest) {
        return authService.register(request.email(), request.name(), request.password(), request.avatarUrl(), extractMeta(httpRequest));
    }

    @PostMapping("/login")
    public Object login(@RequestBody LoginRequest request, HttpServletRequest httpRequest) {
        return authService.login(request.email(), request.password(), extractMeta(httpRequest));
    }

    @PostMapping("/refresh")
    public Object refresh(@RequestBody RefreshRequest request) {
        return authService.refresh(request.refreshToken());
    }

    @PostMapping("/verify-email")
    public Object verifyEmail(@RequestBody TokenRequest request) {
        return authService.verifyEmail(request.token());
    }

    @PostMapping("/resend-verification")
    public Object resendVerification(@RequestBody EmailRequest request) {
        return authService.resendVerification(request.email());
    }

    @PostMapping("/request-password-reset")
    public Object requestPasswordReset(@RequestBody EmailRequest request) {
        return authService.requestPasswordReset(request.email());
    }

    @PostMapping("/reset-password")
    public Object resetPassword(@RequestBody ResetPasswordRequest request) {
        return authService.resetPassword(request.token(), request.newPassword());
    }

    @PostMapping("/confirm-email-change")
    public Object confirmEmailChange(@RequestBody TokenRequest request) {
        return authService.confirmEmailChange(request.token());
    }

    @PostMapping("/logout")
    public Object logout() {
        UUID userId = SecurityUtils.requireUserId();
        return authService.logout(userId);
    }

    @GetMapping("/me")
    public Object me() {
        UUID userId = SecurityUtils.requireUserId();
        return authService.me(userId);
    }

    @PatchMapping("/me")
    public Object updateMe(@RequestBody UpdateProfileRequest request) {
        UUID userId = SecurityUtils.requireUserId();
        return authService.updateProfile(userId, request.name(), request.avatarUrl());
    }

    @GetMapping("/sessions")
    public Object sessions() {
        UUID userId = SecurityUtils.requireUserId();
        return authService.listSessions(userId);
    }

    @DeleteMapping("/sessions/{id}")
    public Object revokeSession(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        return authService.revokeSession(userId, id);
    }

    @PostMapping("/change-password")
    public Object changePassword(@RequestBody ChangePasswordRequest request) {
        UUID userId = SecurityUtils.requireUserId();
        return authService.changePassword(userId, request.currentPassword(), request.newPassword());
    }

    @PostMapping("/request-email-change")
    public Object requestEmailChange(@RequestBody RequestEmailChangeRequest request) {
        UUID userId = SecurityUtils.requireUserId();
        return authService.requestEmailChange(userId, request.newEmail());
    }

    @PostMapping("/delete-account")
    public Object deleteAccount(@RequestBody DeleteAccountRequest request) {
        UUID userId = SecurityUtils.requireUserId();
        return authService.deleteAccount(userId, request.password());
    }

    private SessionMeta extractMeta(HttpServletRequest request) {
        String userAgent = request.getHeader("user-agent");
        String forwarded = request.getHeader("x-forwarded-for");
        String ipAddress = forwarded != null ? forwarded.split(",")[0].trim() : request.getRemoteAddr();
        return new SessionMeta(userAgent, ipAddress);
    }

    public record RegisterRequest(String email, String name, String password, String avatarUrl) {}
    public record LoginRequest(String email, String password) {}
    public record RefreshRequest(String refreshToken) {}
    public record TokenRequest(String token) {}
    public record EmailRequest(String email) {}
    public record ResetPasswordRequest(String token, String newPassword) {}
    public record ChangePasswordRequest(String currentPassword, String newPassword) {}
    public record RequestEmailChangeRequest(String newEmail) {}
    public record UpdateProfileRequest(String name, String avatarUrl) {}
    public record DeleteAccountRequest(String password) {}
}
